evm/chains/sharding/mainchain_handler/vmc_handler.py

```diff
@@ -0,0 +1,240 @@
+import logging
+
+from eth_utils import (
+    to_checksum_address,
+    to_dict,
+)
+
+from evm.utils.hexadecimal import (
+    decode_hex,
+)
+
+from evm.chains.sharding.mainchain_handler.config import (
+    DEPOSIT_SIZE,
+    GASPRICE,
+    TX_GAS,
+)
+
+from evm.chains.sharding.mainchain_handler.vmc_utils import (
+    create_vmc_tx,
+    get_contract_address_from_contract_tx,
+    get_vmc_json,
+)
+
+
+class VMCHandler:
+
+    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMCHandler")
+
+    def __init__(self, mainchain_handler, primary_addr, TransactionClass):
+        """
+        :param primary_addr: address in bytes
+        """
+        self.mainchain_handler = mainchain_handler
+        self.primary_addr = primary_addr
+        self.init_vmc_attributes(TransactionClass)
+        self.setup_vmc_instance()
+
+    def init_vmc_attributes(self, TransactionClass):
+        vmc_tx = create_vmc_tx(TransactionClass, gasprice=GASPRICE)
+        self._vmc_sender_addr = vmc_tx.sender
+        self._vmc_addr = get_contract_address_from_contract_tx(vmc_tx)
+        vmc_json = get_vmc_json()
+        self._vmc_abi = vmc_json['abi']
+        self._vmc_bytecode = vmc_json['bytecode']
+        self.logger.debug("vmc_addr=%s", self._vmc_addr)
+        self.logger.debug("vmc_sender_addr=%s", self._vmc_sender_addr)
+
+    def setup_vmc_instance(self):
+        self._vmc = self.mainchain_handler.contract(
+            to_checksum_address(self._vmc_addr),
+            self._vmc_abi,
+            self._vmc_bytecode,
+        )
+
+    # vmc utils #####################################################
+
+    def get_vmc(self):
+        """
+        :return: web3.eth.contract instance of vmc
+        """
+        return self._vmc
+
+    @to_dict
+    def _mk_contract_tx_detail(self,
+                               sender_addr=None,
+                               gas=TX_GAS,
+                               value=None,
+                               gas_price=None,
+                               data=None):
+        # Both 'from' and 'gas' are required in eth_tester
+        if sender_addr is None:
```

Comment:
This check should potentially be changed to check that `is_canonical_address(sender_addr)`